### ISSUE TITLE
Still No Output From Pylint

### ISSUE THREAD

**Issue by tbpassin** [leo-editor/leo-editor#2586]
See PR #2589 and PR #2590.

With the new PR, the BPM *waits* (via proc.communicate) for all output if BPM.wait is True. 

@tbpassin I expect more work is needed, but I wanted to merge PR #2590 right away. Please test devel with wait set to both True and False.

**Comment by tbpassin** [comment]
Yesterday, I think it was, I did get output on a tiny @file, just a few lines long.  I immediately ran it on VR3 and got no output.

Today, I got no output on a file of about 10 or 15 lines, and got no output.  I also got no output on Freewin.

Leo does not block when I run pylint, and I do get the command name echoed to the console, but not to the log pane.  I never see a message that pylint is finished.

**Comment by tbpassin** [comment]
The problem is definitely in the background process manager.  It's getting to this step at the bottom of the start_process node where it invokes _open_process()_, which is deffed this way::

```
    def open_process():
        return subprocess.Popen(
            command,
            shell=shell,
            stderr=subprocess.PIPE,
            stdout=subprocess.PIPE,
            universal_newlines=True,
        )
```

For this call, _shell_ == False.  But according to
[How to use `subprocess` command with pipes](https://stackoverflow.com/questions/13332268/how-to-use-subprocess-command-with-pipes):

"To use a pipe with the subprocess module, you have to pass shell=True.

However, this isn't really advisable for various reasons, not least of which is security".

I haven't checked the Python docs at this point.

**Comment by edreamleo** [comment]
@tbpassin Looks like a breakthrough, Thomas.

Imo, there are no serious security implications to setting shell = True in the BPM. If someone has access enough to call LeoQtLog.put then they can do anything.

**Comment by edreamleo** [comment]
Arrrgh. I have managed to completely screw everything. My apologies to all.

**Comment by edreamleo** [comment]
PR #2590 sets the wait switch to True. This should ensure that all output is captured, at the cost of hanging Leo while pylint runs.

@tbpassin Let's test this mode to see if it actually work. If so, we can then see if we can solve the hanging problem.

**Comment by tbpassin** [comment]
I agree about the security question.  If attackers could mess with those things, there are more serious and direct bad things they could do without needing to mess wiht Leo.

**Comment by tbpassin** [comment]
What branch is that on?

**Comment by tbpassin** [comment]
I don't see where the output is ever supposed to get out of the pipe and back into Leo to be formatted and displayed.  It should happen here in pylint.run:

```
    for fn, p in data:
        self.run_pylint(fn, p)
    # #1808: return the last data file.
    return data[-1] if data else False
```

It seems to me that either the call that runs pylint should be an async/await kind of operation, or if that's not possible it should be done on a separate thread.  If not, how is control ever going to come back to the right place for further processing of pylint's messages?  Does the POpen process stick the stdout pipe's output into data[-1]?

Confirming the notion that there's no way for the results to come back at the right time and place is that I forced shell = True and still got nothing except a message that pylint had finished.

**Comment by tbpassin** [comment]
This S.O. thread may be very relevant:

[Non Blocking Subprocess,call](https://stackoverflow.com/questions/16071866/non-blocking-subprocess-call)

**Comment by tbpassin** [comment]
Another useful discussion: [PyQt: How to open a program with python without freezing](https://python-forum.io/thread-20474.html)

**Comment by edreamleo** [comment]
@tbpassin Please reopen as necessary.

**Comment by tbpassin** [comment]
With the latest changeset, I do not get any output from pylint unless I set `self.wait = True` in `bpm.start_process()`.  So this changeset will not work for anyone running pylint who does not want to go in and change the value of _self.wait_.

**Comment by tbpassin** [comment]
Why wouldn't this be the Leo-esque way to do it?

pseudo-code:
```
proc = Popen.subprocess(...)
set idle time hook(callback)

def callback:
    check proc returncode
    if returncode:
        cancel hook
        proc.kill()
        outs, errs = proc.communicate()
        compute_msg_string()
        log.put(msg_string)
```

I don't know the mechanics of setting the idle time hook on a callback, and I know you know them very well.  Otherwise I'd have tried it before posting this message.

**Comment by edreamleo** [comment]
@tbpassin Thanks, Thomas, for your interest and work on this issue. Everything works for me, so it's hard to say what is going on. I'm going to ask for more testers in devel.

**Comment by tbpassin** [comment]
I just fetched the latest changeset in devel, and still I get no blocking but not pylint output.

```
Leo 6.6.1-devel, devel branch, build 4bdce3ef48
2022-04-08 10:05:24 -0500
Python 3.9.9, PyQt version 6.2.4
Windows 10 AMD64 (build 10.0.19043) 
```

I see that _wait_ get set to False in bpm.init().  So it's a class variable, not even an instance variable.  Unless bpm is a singleton class, isn't that asking for trouble (not that it's an issue for the pylint problem)?

On my system, bpm.wait = false doesn't give output, True does, with blocking.

**Comment by edreamleo** [comment]
@tbpassin The BPM is a singleton, so I don't think there is a problem.

**Comment by tbpassin** [comment]
Well, progress. The check_process() method polls the pylint process for completion but never gets it:

```
        if self.pid.poll() is None:  # The process is still running.
            pass
```

After I had let it run for about 6000 seconds, there had still been no output and the task was still visible in taskmanager, though shown as using no CPU.  When I manually killed the pylint task with the Windows taskmanager, then the pylint output was printed to the log pane.

**Comment by tbpassin** [comment]
From reading, if the output pipe fills up, the process might always return None when polled.  Not sure how or if this applies here.

**Comment by edreamleo** [comment]
@tbpassin Thanks Thomas. You may have put your finger on the problem. I have a half memory of trying to read the pipe while the process is running, which one might think would cause the pipe not to fill up, but iirc that didn't work.

You've given me something to chew on.

BTW, how much output did you finally see in the log pane?  Can you send me a file that creates that much output?

**Comment by tbpassin** [comment]
I'm running it on vr3.  There are a lot of lines about not being able to find Qt objects - we've seen those before.  I just tricked it into telling me how many characters it's returning by putting a self.pid.stdout.read() in `check_process()`.  That blocked, but it returned 15534 characters for the output up to that point.  So it was at least that long.

```
def check_process(self):
    """Check the running process, and switch if necessary."""
    # #2428: Handle all output only after the process has completed.
    #        There should be no danger of a deadlock because
    #        there is only one running subprocess.
    self.check_count += 1
    if self.pid:
        if self.pid.poll() is None:  # The process is still running.
            if self.check_count % 10 == 0:
                g.es('====', self.check_count, self.pid.returncode)
                print('----', len(self.pid.stdout.read()))
```

I have found lots of complaints online from people who never get anything returned from poll().  The response in all cases is about pipes getting filled up or blocking each other.

I wonder if the output can go to a StringIO instead of a pipe.  I was just about to go looking to see if I can find anything.  I don't see why it couldn't.  That should remove any possibility of filling up the output channel.

**Comment by edreamleo** [comment]
@tbpassin Potential breakthrough. Quite by accident I stumbled upon [this example](https://docs.python.org/3/library/venv.html#an-example-of-extending-envbuilder) from the venv module. It creates reader processes in separate threads.

I'll play with these ideas.

**Comment by edreamleo** [comment]
@tbpassin I'm pretty sure messing with the polling loop won't work.

**Comment by tbpassin** [comment]
I just read that if you use threads here, you tend to end up with zombie threads once the pipe is done.

**Comment by tbpassin** [comment]
I found this on Microsoft docs:

"To write to the pipe, use the pipe's write handle in a call to the [WriteFile](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-writefile) function. The WriteFile call does not return until it has written the specified number of bytes to the pipe or an error occurs. If the pipe buffer is full and there are more bytes to be written, WriteFile does not return until another process reads from the pipe, making more buffer space available. The pipe server specifies the buffer size for the pipe when it calls CreatePipe."

From [Anonymous Pipe Operations](https://docs.microsoft.com/en-us/windows/win32/ipc/anonymous-pipe-operations).

**Comment by tbpassin** [comment]
I tried running devel with my old Python 3.7 version.  Either because it has an older version of pylint, or it used the old rcfile location, it found only one line to warn about, and it completed normally as expected.

I'd say this more or less confirms that it's large outputs that cause the trouble.  Perhaps the default pipe size is capped at 64k in my version of Windows.  I haven't been able to find out anything useful about the default pipe size.

Maybe the easiest way would be to write to an actual file and then read that file.

**Comment by edreamleo** [comment]
@tbpassin Really? Replacing pipes with (temporary) files?

Hmm. Maybe it's a blind spot on my part.  It's worth a try.

**Comment by tbpassin** [comment]
Wait a bit to try it.  I may possibly have a solution that doesn't need files.  Let me fuss with it a bit more first.

**Comment by tbpassin** [comment]
I think I have a solution.  In the check_process loop, I put in a readlines(16).  That only blocks until pylint starts pumping output. read() or readlines() would block, but with a size hint, they don't need to wait for more output.

I collect all those `readlines(16)` lines and combine them with the lines gotten after poll() returns.  "16" seems to be a good number because pylint never seems to put out more than one line per checking interval, so nothing gets discarded.

self.lines = [] needs to be added to the class __init__(), and it needs to be re-initialized before every run of pylint.  I'm not showing that here.

The results on VR3 seem to be as expected and I'm fairly sure nothing is getting lost.

```
def check_process(self):
    """Check the running process, and switch if necessary."""
    # #2428: Handle all output only after the process has completed.
    #        There should be no danger of a deadlock because
    #        there is only one running subprocess.
    if self.pid:
        if self.pid.poll() is None:  # The process is still running.
            if self.check_count %5 == 0:
                self.lines.extend(self.pid.stdout.readlines(16))
                print('===', len(self.lines))  # Only for experimenting, remove later
            pass
        else:  # The process has completed.
            if self.wait:
                pass
            else:
                self.lines.extend(self.pid.stdout)
                for s in self.lines:
                    self.data.number_of_lines += 1
                    self.put_log(s)
            self.end()  # End this process.
            self.start_next()  # Start the next process.
    elif self.process_queue:
        self.start_next()  # Start the next process.
```

**Comment by edreamleo** [comment]
@tbpassin Hmm. But what if there is no output. Now isn't pylint going to block Leo until pylint finishes?

**Comment by tbpassin** [comment]
Well, yes, I suppose it would.  I didn't find any large files in the Leo codebase that didn't generate lots of pylint warnings, so I went to vr3 and disabled all the warning type that were coming up.  There were only four of them.

And yes, it mostly blocked, with a few short periods of not being blocked.  In addition, I imagine that pylint finished more quickly since it didn't have to check for those particular conditions.

Here is a link that looks right on target.  It's about using a separate thread to read the Popen pipe to avoid deadlock when it's full:

[Python trick: asynchronously reading subprocess pipes](https://www.stefaanlippens.net/python-asynchronous-subprocess-pipe-reading/)

**Comment by edreamleo** [comment]
@tbpassin Good find, Thomas. The link you provide looks similar to the code in the venv docs.

I'd like to play with code, unless you want to claim the "first play" rights by virtue of you having found the code.

**Comment by tbpassin** [comment]
No, please go right ahead.  I think this is really your baby, since you tried to get non-blocking pylint working in the first place.  It seems to me that the only use case would be if one wanted to run a series of external files through pylint - I at any rate don't mind very much waiting the minute or so of blocking for any one file.  Your machinery is set up to handle a series of pylint or other checker commands, and you are probably the only one who knows that.  I can see why one would want non-blocking pylint in that case.

My main interest has been to get pylint output, whether blocking or not.  And I don't like it when my tools don't quite do what they are supposed to - it makes me happy when I can help do something about that, and with Leo it sometimes is feasible.

I've also seen subprocess code using async/await, but I'm not sure if it addresses our particular problem or not.

**Comment by edreamleo** [comment]
@tbpassin OK, I'll forge ahead. Not sure about async/await.

**Comment by tbpassin** [comment]
I can see the bare bones of how to do it with threads.  The problem comes about in check_process.  If we do nothing, the pipe may fill up and stall.  If we read it to prevent stalling, then Leo may block.  So we just have the thread do what check_process does.  I would think that it could set up its own timer and callback, and that could also block for all we care.

When the pylint process finishes, the thread could put the results to the log pane, and that would eliminate needing to signal the main code line about the completion and pass the results to it.  To make this work, the thread would need a reference to g.  I don't even think it would need a reference to the bpm.  The thread would clean up its timer before joining.

Well, maybe the thread would need a reference to the bpm after all.  The bpm needs to know when the run is done so it can start another.

**Comment by edreamleo** [comment]
I think I agree with most of what you are saying. I have a plan.  We'll see how it works.

**Comment by edreamleo** [comment]
@tbpassin Please see PR #2592. The new code is dead simple.  The Aha, we can call proc.communicate *after* the process has finished. This works for me. There should be no danger of hanging Leo.

The question is, can the process ever block?  I'm not sure why it would, but we shall see.

**Note**: I *really* do not want add a new user setting telling whether to call proc.communicate immediately, which would hang Leo.

**Comment by tbpassin** [comment]
This can't work.  The pipe will still fill up and stall the pylint process. It was one of the first things I tried yesterday, actually.  I've just verified this running your new devel branch on vr3.  I never get the output, and the new pylint process stalls and hangs around using no cpu, as shown by the Windows task manager.  As soon as I killed the pylint process with task manager, (some of) the output appeared in the log pane.  What appeared must have been how much was in the pipe.  A lot was missing.

I think you need a test case that will produce lots of output.  Can you change your pylint rc file to check for I1101, "c-extension-no-member"?  That will produce enough output from vr3 to let you test things yourself.

**Comment by edreamleo** [comment]
As I said in the discussion group, I now have a proper test. Imo, there is no simple solution to blocked pipes, except to call proc.communicate once, just after creating the separate process.

So I have given up, for now, and set wait=True. This appears to work.

Aha: there is an entirely new approach: just run pylint (or mypy) *outside* of Leo, and then find some way to copy the output to the log pane so that the links become clickable! Clearly, this can be made to work!

**Comment by tbpassin** [comment]
Yes, I thought of that too, and it does seem workable.  I was saving it in case we came to another solution.  It should probably be run using POpen or it will probably end up blocking too.

**Comment by tbpassin** [comment]
OK, I have a working, non-blocking threaded solution.  The POpen process is checked in a separate thread.  When this thread is done, and the pylint output collected, it stores the collected lines in a bpm variable.  On the timer thread, the called task checks the variable to see if it is not [].  If it has contents, then it puts those contents to the log pane.  Writes to and reads from the variable are protected by a lock so they are guaranteed to be atomic.

Here is the thread procedure:

```
def thrd_pipe_proc(self):
    """The threaded procedure to handle the Popen pipe.

    When the process has finished, its output data is
    collected and placed into bpm.process_return_data.
    A timer callback in bpm checks to see if this data has 
    arrived and if so, uses it.
    """
    lines = []
    while self.pid and not self.pid.poll():
        # Prevent buffer from filling, which would cause blocking
        lines.extend(self.pid.stdout.readlines(16))
        print('===', len(lines))  # Only for experimenting, remove later

    if self.pid:
        lines.extend(self.pid.stdout)
    while self.lock.locked():
        sleep(0.03)
    self.lock.acquire()
    self.process_return_data = lines
    self.lock.release()
```
Here is the timer callback:

```
def on_idle(self):
    """The idle-time callback for leo.commands.checkerCommands."""
    try:
        g.app.gui.qtApp.processEvents()  # #2528.
    except Exception:
        pass

    if self.process_return_data:
        # Wait for data to be fully written
        while self.lock.locked():
            sleep(0.03)
        for s in self.process_return_data:
            self.data.number_of_lines += 1
            self.put_log(s)

        self.end()  # End this process.
        self.start_next()  # Start the next process.
```
Sleeps in the thread cause blocking in the main GUI, so you want to keep any sleeps very short.

There's some setup required, of course.  Here's part of it:

```
def start_process(self, c, command, kind,
    fn=None,
    link_pattern=None,  # None, string, or re.pattern.
    link_root=None,
):
    """
      def open_process():
        proc = subprocess.Popen(
            command,
            shell=False,  # #2586
            stderr=subprocess.PIPE,
            stdout=subprocess.PIPE,
            universal_newlines=True,
        )
        return proc

    def start_timer():  # #2528 & #2557.
        if self.wait:
            return
        if not self.timer.isActive():
            self.timer.start(100)
        thread.start_new_thread(self.thrd_pipe_proc, ())

    self.process_return_data = []
    # ... Carry on with existing code
```

There is some cleanup needed, and some code review to make sure I didn't drop out some functionality by mistake.  And some more testing by me, of course. 
 Probably tomorrow I'll be able to put in a PR so others can try it out.

**Comment by edreamleo** [comment]
@tbpassin When I said "just run pylint (or mypy) outside of Leo" I was thinking "run pylint (or mypy) as usual, from a .cmd file. These never block.

**Comment by edreamleo** [comment]
@tbpassin Congratulations. You may have rescued the BPM. I'll look forward to your PR.

The acid test will be the responsiveness of Leo. If it's acceptable, I'll accept the PR. Otherwise, I'll deprecate the BPM.

I have just created #2593, which suggests that copying text into Leo's log window will automagically create clickable links.  Devs could run (using scripts) pylint or mypy in separate consoles, with *zero* impact on Leo. If errors appear, devs would paste the links into Leo's log window.  Leo would never wait!

*Note*: While considering whether to deprecate the BPM, searches revealed several functions that can (should!) be removed. So all this work is having knock-on benefits.

**Comment by tbpassin** [comment]
Was this the intended work flow for the bpm?

1. User starts pylint on one file;
2. While that task is still running, user starts pylint on a second file;
3. repeat for more files.

I'm not sure how helpful that would be, since the results for a previous run would scroll up out of sight when the later ones come in.  In my experience, one needs to spend time - maybe a lot of time - checking out pylint results before moving on, and it's not very workable to try to scroll back to the right place in the log pane output to try to do that.

If the pylint command were blocking, at least that would remove that potential issue.  OTOH, I always dislike blocking if it can be avoided.  But if the command does block, then the user might overlap pylint commands and that has to be provided for somehow.

**Comment by edreamleo** [comment]
@tbpassin Yes, that was *one* intended workflow. I might use this when not expecting any complaints from pylint. The queue-related code in the BPM is intended to honor various delayed requests. But your innovations are paramount.

**Comment by tbpassin** [comment]
I've implemented the changes and they work like a charm.  I'd like you to test the branch on my fork, and also do some code review.  This changeset has some diagnostic print statements which I'd remove before a PR.

I am able to run pylint on every file in the Qt-only plugins tree and continue editing and navigating in Leo while they are all getting checked (with lots of pylint messages), and Leo feels normal while I do it.

I found one anomaly that I took care of.  If pylint finished with no warning messages, the checking loop never terminated.  The current code monitors for this condition and breaks out of the loop if necessary.

I have tried to make minimal changes to the bpm code, but I did remove those lines in bpm.end() that duplicated the output to the log pane.  I also suppressed printing the pylint output to the console.  I really dislike that: it clutters up the console, it's nearly unreadable and very hard to do anything with, and it duplicates what displays in the log pane.  I left the commented line of code in to make it easy to restore if you just can't live without it.

My fork is at [tbp's github fork](https://github.com/tbpassin/leo-editor).  The branch is tbp-pylint.

Here are the tests I've been doing.

1. Run pylint against vr3.
2. Run pylint against plugins/colorize_headlines.py (this is the file that produced no warning messages and stalled earlier versions of the code).
3. Start pylint on vr3, and while it is running, start it (from the minibuffer) on vr.
4. Select the _Qt only plugins_ top node, and run pylint.  All files in the tree should get pylinted successfully.

Once you agree, I'll remove the diagnostic print statements and issue a PR.

**Comment by tbpassin** [comment]
You may want to check my implementation with mypy.  I have only ever checked it with pylint.

**Comment by tbpassin** [comment]
Turns out that I see stalled loops sometimes for files that do have pylint warnings.  I don't really know why that is.  But the loop breakout mechanism handles those cases perfectly, so I don't feel inclined to go digging into it any further.

**Comment by tbpassin** [comment]
I considered and discarded two design choices for the threaded process checker:

1.  Having the threaded process start the actual POpen process;
2.  Having the threaded process put the results to the log pane.

I rejected those ideas because I thought the thread should do only one thing, which is to check for completion.  I knew how to pass the results back to the bpm, so there was no need to have the threaded process do the output.  And anyway, the bpm is responsible for just what to do with the output, it's not the business of the process checker thread.

**Comment by tbpassin** [comment]
Ha!  I used  `if not self.pid.poll():`  but I should have used `if self.pid.poll() is not None:`.  Making this change cleared up all those cases where we seemed to be missing the poll() change.  I'll leave the protective loop-breakout code in place as a backup, but it doesn't seem to ever get hit any more.

**Comment by tbpassin** [comment]
Oops, that should have been `if self.pid.poll() is None`!.  This way we get the previous behavior with the pylint output coming in during the loop.  Aparently with the test inadvertently reversed the loop was skipped but we still waited for the task to be finished in the line 

`        lines.extend(self.pid.stdout)`

This, however, didn't stall with a full pipe.  Why not, I don't know since it did before we split this code out into the separate thread.

Well, anyway, it's working as intended.

**Comment by tbpassin** [comment]
Doh! This is the perfect place to use communicate().  It will block the thread but that's why we created the thread in the first place.  So the whole thread function becomes only this:

```
def thrd_pipe_proc(self):
    """The threaded procedure to handle the Popen pipe.

    When the process has finished, its output data is
    collected and placed into bpm.process_return_data.
    An on_idle task in bpm checks to see if this data has 
    arrived and if so, uses it.
    """
    result, err = self.pid.communicate()
    while self.lock.locked():
        sleep(0.03)
    self.lock.acquire()
    self.process_return_data = result.split('\n')
    self.lock.release()
```
It works perfectly.

Please try this version in my github fork.  The changeset is

`e55c0b27c9479868a07439c5cb21fde32cbcf418`

There is still some blocking, but it only occurs while the pylint output is actually being put to the log file.  I'd say we can live with that!

**Comment by edreamleo** [comment]
@tbpassin Looks like great work. I'll test it myself today.

**Comment by edreamleo** [comment]
@tbpassin Just tested the code. It seems to work. Please do create a PR.

A cleanup.  In ekr-hang5, check_processed has been "flattened" as follows:
```python
check_count = 0

def check_process(self):
    """Check the running process, and switch if necessary."""
    self.check_count += 1
    # g.trace(self.check_count)
    if self.pid and self.pid.poll() is None:
        return
    if self.pid:
        self.end()  # End this process.
    self.start_next()  # Start the next process.
```
I would like you to use this as a "template". So I'm *guessing* that would give you:
```python
check_count = 0

def check_process(self):
    """Check the running process, and switch if necessary."""
    self.check_count += 1
    # g.trace(self.check_count)
    if self.pid and self.pid.poll() is None:
        return
    if self.pid:
        for s in self.pid.stdout:
            self.put_log(s)
        self.end()  # End this process.
    self.start_next()  # Start the next process.
```
Another cleanup.  I have deleted data.number_of_lines everywhere.

Finally, don't worry about mypy. I can always tweak that after merging the PR.

**Comment by tbpassin** [comment]
I actually don't use check_process at all and will remove it for the PR.  The threaded equivalent is now so simple that I don't think anything can be taken away.  But I'll look and see what I can apply elsewhere.

**Comment by edreamleo** [comment]
@tbpassin Even better!

**Comment by edreamleo** [comment]
Closed via PR #2595.

**Comment by edreamleo** [comment]
On Sat, Apr 9, 2022 at 4:31 PM Thomas Passin ***@***.***>
wrote:

> This can't work. The pipe will still fill up and stall the pylint process.
>
OK. I do now have a proper test file.  I added a whole bunch of lines like:

    a = 1; print(a)

to the VR3 ctor.  Now I do see the hang.

The following works for me.  check_process calls communicate in the polling
loop.  Like this:

    if self.pid and self.pid.poll() is None:
        # The process is still running.
        try:
            outs, errs = self.pid.communicate(timeout=0.1)
            for s in g.splitLines(outs):
                self.put_log(s)
        except subprocess.TimeoutExpired:
            pass
        return

But with this change I got duplicates of messages! To fix this,
check_process should *not* call communicate when the process is complete.
That is, check_process ends like this:

    if self.pid:
        self.end()  # End this process.
    self.start_next()  # Start the next process.

This now works for me on the substantial test.  I have just updated the PR
accordingly.  How does it work for you???

BTW, the repeated calls to communicate reduce Leo's responsiveness
slightly, but not by much.

Edward

**Comment by tbpassin** [comment]
Wow, I'm really cold on this one so it will take me some time to get my mind around it all.  I'm puzzled because what you are describing doesn't seem like what I (vaguely) remember.  I thought we got rid of check_process, and in fact I don't see it in the code at all (I haven't updated from the devel branch for a week or so).

At a quick look, the only things I see to change (from my original version) would be the wait loops before lock.acquire(). I have since learned that lock.acquire() will wait until the lock is available, so those loops aren't needed.

"repeated calls to communicate reduce Leo's responsiveness" - in the code we've been using since April, there is only one call to communicate() per background process.  It blocks the background thread (not Leo itself) until the background process is finished, but that's as designed and is the reason it's on a separate thread.  Since (as I remember it) the BPM won't launch the next process from its queue until the first one returns, this should never cause any problem.  The only activity that gets repeated by Leo is an on-idle check to see if the results are back from the background process.  And that's just a check to see if the list `self.process_return_data` is non-empty or not. This check would not incur a noticeable delay.

I get the impression that your remarks may apply to a version of the BPM from before my April PR.  Would you mind checking that, please?  Could you perhaps have gone back to your old BPM branch from April but never updated it to the current code?

What branch are your current changes on?  I hope it's not devel : ).

**Comment by tbpassin** [comment]
As I remember it, we had a problem with `communicate()` because it would not always return all the data generated by the subprocess.  It was as if the pipe has a very limited capacity.  I see from the Python docs:

"Note The data read is buffered in memory, so do not use this method if the data size is large or unlimited."

Also from the docs:

"If the process does not terminate after timeout seconds, a [TimeoutExpired](https://docs.python.org/3/library/subprocess.html#subprocess.TimeoutExpired) exception will be raised. Catching this exception and retrying communication will not lose any output.

The child process is not killed if the timeout expires, so in order to cleanup properly a well-behaved application should kill the child process and finish communication:

```
proc = subprocess.Popen(...)
try:
    outs, errs = proc.communicate(timeout=15)
except TimeoutExpired:
    proc.kill()
    outs, errs = proc.communicate()
```
"

It seems to me that a change from my design - of calling `communicate()` just once on the process's thread, where it can block as long as it needs to - to some other that uses `communicate()` on the main thread repeatedly would entail a lot more complexity for no gain (and potentially reduce Leo's responsiveness).

**Comment by edreamleo** [comment]
@tbpassin Hi Thomas. Thanks for taking a look. It would be nice to get this resolved.
